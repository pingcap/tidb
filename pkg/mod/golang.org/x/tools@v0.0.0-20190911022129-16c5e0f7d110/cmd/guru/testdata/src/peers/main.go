package main

// Tests of channel 'peers' query.
// See go.tools/guru/guru_test.go for explanation.
// See peers.golden for expected query results.

var a2 int

func main() {
	chA := make(chan *int)
	a1 := 1
	chA <- &a1

	chA2 := make(chan *int, 2)
	if a2 == 0 {
		chA = chA2
	}

	chB := make(chan *int)
	b := 3
	chB <- &b

	<-chA  // @pointsto pointsto-chA "chA"
	<-chA2 // @pointsto pointsto-chA2 "chA2"
	<-chB  // @pointsto pointsto-chB "chB"

	select {
	case rA := <-chA: // @peers peer-recv-chA "<-"
		_ = rA // @pointsto pointsto-rA "rA"
	case rB := <-chB: // @peers peer-recv-chB "<-"
		_ = rB // @pointsto pointsto-rB "rB"

	case <-chA: // @peers peer-recv-chA' "<-"

	case chA2 <- &a2: // @peers peer-send-chA' "<-"
	}

	for range chA {
	}

	close(chA) // @peers peer-close-chA "chA"

	chC := make(chan *int)
	(close)(chC) // @peers peer-close-chC "chC"

	close := func(ch chan *int) chan *int {
		return ch
	}

	close(chC) <- &b // @peers peer-send-chC "chC"
	<-close(chC)     // @peers peer-recv-chC "chC"
}
