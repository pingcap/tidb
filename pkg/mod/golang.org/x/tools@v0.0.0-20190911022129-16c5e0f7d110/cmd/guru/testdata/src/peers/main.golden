-------- @pointsto pointsto-chA --------
this chan *int may point to these objects:
	makechan
	makechan

-------- @pointsto pointsto-chA2 --------
this chan *int may point to these objects:
	makechan

-------- @pointsto pointsto-chB --------
this chan *int may point to these objects:
	makechan

-------- @peers peer-recv-chA --------
This channel of type chan *int may be:
	allocated here
	allocated here
	sent to, here
	sent to, here
	received from, here
	received from, here
	received from, here
	received from, here
	received from, here
	closed, here

-------- @pointsto pointsto-rA --------
this *int may point to these objects:
	peers.a2
	a1

-------- @peers peer-recv-chB --------
This channel of type chan *int may be:
	allocated here
	sent to, here
	received from, here
	received from, here

-------- @pointsto pointsto-rB --------
this *int may point to these objects:
	b

-------- @peers peer-recv-chA' --------
This channel of type chan *int may be:
	allocated here
	allocated here
	sent to, here
	sent to, here
	received from, here
	received from, here
	received from, here
	received from, here
	received from, here
	closed, here

-------- @peers peer-send-chA' --------
This channel of type chan *int may be:
	allocated here
	sent to, here
	received from, here
	received from, here
	received from, here
	received from, here
	received from, here
	closed, here

-------- @peers peer-close-chA --------
This channel of type chan *int may be:
	allocated here
	allocated here
	sent to, here
	sent to, here
	received from, here
	received from, here
	received from, here
	received from, here
	received from, here
	closed, here

-------- @peers peer-close-chC --------
This channel of type chan *int may be:
	allocated here
	sent to, here
	received from, here
	closed, here

-------- @peers peer-send-chC --------
This channel of type chan *int may be:
	allocated here
	sent to, here
	received from, here
	closed, here

-------- @peers peer-recv-chC --------
This channel of type chan *int may be:
	allocated here
	sent to, here
	received from, here
	closed, here

